src/main/java/com/google/devtools/build/lib/bazel/repository/downloader/DownloadManager.java

```diff
@@ -71,9 +72,11 @@ public interface CredentialFactory {
     Credentials create(Map<URI, Map<String, List<String>>> authHeaders);
   }
 
-  public DownloadManager(RepositoryCache repositoryCache, Downloader downloader) {
+  public DownloadManager(
+      RepositoryCache repositoryCache, Downloader downloader, HttpDownloader httpDownloader) {
```

Comment:
What's the difference between Downloader and HttpDownloader? Can you please add a comment somewhere to explain why we need both?

Reply:
I documented the need for the separate `HttpDownloader` in the constructor, and renamed the field to be clearer. As to why exactly the bzlmod downloads want to side-step the `GrpcRemoteDownloader`, if set up, I don't know. This is just semantics-preserving. @fmeum I guess this is your department?

If we don't actually need this special case, we could probably unify the two download methods more by an internal method that gets an argument whether to check the is-downloads-disabled-flag, and by taking some sort of outputstream - which in the bzlmod case is an in-memory-buffer, and in the general case is backed by the FS. That's a future refactoring though, that's unrelated to this PR.

Reply:
Thanks a lot! I believe the reason we use a different one for Bazel registry is because we don't want `--repository_disable_download` to stop Bazel from registries?

https://github.com/bazelbuild/bazel/pull/23106/files#diff-5d6816db4205f516688ad1a3f96d88ccdc0b03523a1658621ede92589268b3baL358

And that reminds me to ask does `--repository_disable_download` have any effect on downloading remote build artifacts? /cc @fmeum

Reply:
I don't know enough about the use cases of `GrpcRemoteDownloader` to decide whether we could expect it to have access to Bazel registries, but I would guess that this should work and it was only left out of the Bzlmod implementation because nobody was certain that it would work. @Wyverald Is that correct? :-)

I'm pretty sure `--repository_disable_download` isn't related to this since I only very recently made it so that this flag doesn't stop registry access. I'm also pretty sure that it doesn't interact in any way with remote build artifact downloads.

Reply:
Ah, I think I got it:
* `repository_disable_download` is checked in `DownloadManager`, so it doesn't matter to this discussion (but it explains why we need two different top-level methods here)
* The method `downloadAndReadOneUrl` that's used by the bzlmod downloader only exists on `HttpDownloader`, not on the `Downloader` interface or the `GrpcRemoteDownloader`.

I suppose some refactoring could unify this interface more, and then, potentially, also access the registry through the `GrpcRemoteDownloader`. Then we'd be able to get rid of the separately passed in `HttpDownloader`.

Remote build artifact downloading (through HTTP) goes through `HttpCacheClient` which is an entirely different code path, and doesn't use `HttpDownloader` - it rather uses some low-level netty API for implementing the file transfer.

Reply:
After resolving the other comment, now I believe we can actually merge those two, it should be fine to use GrpcRemoteDownloader for downloading files from Bazel registry. Let me know if you want to merge them in this PR or a follow up one.

Reply:
I'd prefer to get this PR in, so we can get the backport in before the 7.3.0rc1 deadline (Monday).
Merging those methods does require thinking about the API, as the current `Downloader` API takes multiple URLs and an output dir (I think), whereas the bzlmod download supplies a single URL and expects `byte[]` in return.

Reply:
SG, I'll get this one merged! Thanks for the great work!